### PR TITLE
Import-DscResource should allow to overwrite DSC built-in resources.

### PR DESCRIPTION
**Pull request by amitsara**
# PR Summary

**Porting changes from WMF 5.0**

**Issue:**
    User cannot use PsDscResources in DSC configuration and compile it on PowerShellCore (Windows). 
    **For example:** Following configuration will fail on PowerShellCore (Windows) because of resource name conflicts with DSC Built-In resource.

Configuration WinRMService
{
    Import-DscResource -ModuleName GuestConfiguration -ModuleVersion 1.12.0.0

    Service winrmService {
        Name = 'winrm'
        State = 'Running'
    }
}
WinRMService

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)

### CODE REVIEW

**Review by zjalali — APPROVED** [review]
:shipit:

**Comment by TravisEz13** [comment]
@PoshChan Please retry all

**Comment by PoshChan** [comment]
@TravisEz13, successfully started retry of `PowerShell-CI-static-analysis, PowerShell-CI-Windows, PowerShell-CI-macOS, PowerShell-CI-Linux`

**Comment by TravisEz13** [comment]
@PoshChan Please rerun all

**Comment by PoshChan** [comment]
@TravisEz13, successfully started rebuild of `PowerShell-CI-static-analysis, PowerShell-CI-Windows, PowerShell-CI-macOS, PowerShell-CI-Linux`

**Comment by TravisEz13** [comment]
@PoshChan Please retry macos

**Comment by PoshChan** [comment]
@TravisEz13, successfully started retry of `PowerShell-CI-macOS`

**Comment by TravisEz13** [comment]
@PoshChan Please rerun windows

**Comment by PoshChan** [comment]
@TravisEz13, successfully started rebuild of `PowerShell-CI-Windows`

**Comment by TravisEz13** [comment]
@PoshChan please rerun macos

**Comment by PoshChan** [comment]
@TravisEz13, successfully started rebuild of `PowerShell-CI-macOS`

File: src/System.Management.Automation/DscSupport/CimDSCParser.cs [hunk 1]
```diff
@@ -3221,6 +3283,21 @@ public static bool ImportCimKeywordsFromModule(PSModuleInfo module, string resou
 
             return false;
         }
+        
+        /// <summary>
+        /// Clear the 'IsImportedImplicitly' flag when explicitly importing a resource
+        /// </summary>
+        /// <param name="module"></param>
+        /// <param name="cimClass"></param>
+        public static void ClearImplicitlyImportedFlagFromResourceInClassCache(PSModuleInfo module, CimClass cimClass)
```
Comment (TravisEz13) [inline src/System.Management.Automation/DscSupport/CimDSCParser.cs]: This is a public method on an internal class so it is not public to every one?

**Review by TravisEz13 — COMMENTED** [review]
see comments

**Comment by TravisEz13** [comment]
@PoshChan Please rerun windows

**Comment by PoshChan** [comment]
@TravisEz13, successfully started rebuild of `PowerShell-CI-Windows`

**Comment by TravisEz13** [comment]
@SteveL-MSFT What level of testing do we want for this in this repo?

**Comment by SteveL-MSFT** [comment]
@amitsara when do you expect to submit a PR of tests?  We need sufficient tests to ensure this continues to work in the future.

**Comment by TravisEz13** [comment]
@PoshChan Please rerun windows

**Comment by PoshChan** [comment]
@TravisEz13, successfully started rebuild of `PowerShell-CI-Windows`

**Comment by adityapatwardhan** [comment]
Closing and reopening PR to re-trigger Windows CI.

**Comment by TravisEz13** [comment]
@amitsara  Can your respond to https://github.com/PowerShell/PowerShell/pull/9879#issuecomment-501897997

File: src/System.Management.Automation/DscSupport/CimDSCParser.cs (hunk 1 above)
Comment (amitsara) [reply src/System.Management.Automation/DscSupport/CimDSCParser.cs]: I will make it private, its not being used outside of this class.

---
In reply to: [293527097](https://github.com/PowerShell/PowerShell/pull/9879#discussion_r293527097) [](ancestors = 293527097)

**Comment by amitsara** [comment]
We already have tests in PowerShell repo that makes sure that DSC resource can be imported and Configuration can be compiled on Linux and Windows machines. There were some issues related to test data that I fixed in my last check-in. I think we have sufficient coverage in PowerShell repo.

---
In reply to: [501897997](https://github.com/PowerShell/PowerShell/pull/9879#issuecomment-501897997) [](ancestors = 501897997)

**Pull request merged**